src/core/runtime/amd_aql_queue.cpp

```diff
@@ -479,6 +479,43 @@ uint32_t AqlQueue::ComputeRingBufferMaxPkts() {
   return uint32_t(max_bytes / sizeof(core::AqlPacket));
 }
 
+void AqlQueue::CloseRingBufferFD(const char *ring_buf_shm_path, int fd) const {
+#if !defined(HAVE_MEMFD_CREATE)
+  shm_unlink(ring_buf_shm_path);
+#endif
+  close(fd);
+}
+
+int AqlQueue::CreateRingBufferFD(const char *ring_buf_shm_path,
+                                 uint32_t ring_buf_phys_size_bytes) const {
+
+  int fd;
+#ifdef HAVE_MEMFD_CREATE
+  fd = syscall(__NR_memfd_create, ring_buf_shm_path, 0);
+
+  if (fd == -1)
+    return -1;
+
+  if (ftruncate(fd, ring_buf_phys_size_bytes) == -1) {
+    CloseRingBufferFD(ring_buf_shm_path, fd);
+    return -1;
+  }
+#else
+  fd = shm_open(ring_buf_shm_path, O_CREAT | O_RDWR | O_EXCL,
+                S_IRUSR | S_IWUSR);
+
+  if (fd == -1)
+    return -1;
+
+  if (posix_fallocate(fd, 0, ring_buf_phys_size_bytes) != 0)
+    CloseRingBufferFD(ring_buf_shm_path, fd);
+    return -1;
+  }
+#endif
+
+  return fd;
+}
+
 void AqlQueue::AllocRegisteredRingBuffer(uint32_t queue_size_pkts) {
   if (agent_->profile() == HSA_PROFILE_FULL) {
```

Comment:
Tom, could you set this branch to false and try it?  It may be that the dGPU driver could set up this mapping and simplify this a lot.  I'd try it but I'm lacking an APU and a RedHat install.  Thanks.

Reply:
Do you mean this:

-  if (agent_->profile() == HSA_PROFILE_FULL) {
+ if (false) {